### PR TITLE
add support of pkcs11 uri using new openUrl function

### PR DESCRIPTION
**Pull request by JimmyHamel**
Created a new function for the session object called openUrl which has some support to retrieve an object by parsing a pkcs11 url.

### CODE REVIEW

**Comment by JanSlabon** [comment]
While I love this method, I'm not sure if it is the right name for it?

Shouldn't it be named e.g. `findObjectsByUri()` to be more aligned to the other `findObjects()` method? Regarding `Url` vs. `Uri` I'm a bit unsure... In [RFC 7512](https://datatracker.ietf.org/doc/html/rfc7512) it is called `URI` throughout.

**Comment by vjardin** [comment]
+1 with @gamringer

File: pkcs11session.c [hunk 1]
```diff
@@ -736,6 +701,166 @@ void pkcs11_session_shutdown(pkcs11_session_object *obj) {
...
+
+    pkcs11_session_object *objval = Z_PKCS11_SESSION_P(ZEND_THIS);
+
+    array_init(return_value);
+    pkcs11_find_objects(objval, tmpl, nbAttributes, return_value);
+
+    freeTemplate(tmpl);
+}
+
+void pkcs11_find_objects(pkcs11_session_object *objval,  CK_ATTRIBUTE *tmpl, int nbAttributes, zval * return_value) {
```
Comment (vjardin) [inline pkcs11session.c]: why void ?

it should be a static function.

File: pkcs11session.c [hunk 2]
```diff
@@ -736,6 +701,166 @@ void pkcs11_session_shutdown(pkcs11_session_object *obj) {
...
+        pkcs11_object_object* object_obj;
+        object_init_ex(&zp11objectobj, ce_Pkcs11_P11Object);
+        object_obj = Z_PKCS11_OBJECT_P(&zp11objectobj);
+        object_obj->session = objval;
+        object_obj->object = hObject;
+        GC_ADDREF(&objval->std);
+        zend_hash_next_index_insert(Z_ARRVAL_P(return_value), &zp11objectobj);
+    }
+
+    rv = objval->pkcs11->functionList->C_FindObjectsFinal(objval->session);
```
Comment (vjardin) [inline pkcs11session.c:866]: what is `rv` used for ?

**Review by vjardin — COMMENTED** [review]
please, can you apply some cosmetic changes ?

File: pkcs11session.c (hunk 2 above)
Comment (gamringer) [reply pkcs11session.c:866]: rv should be returned and checked in the calling method to throw an exception if not CKR_OK, as it is everywhere else

File: pkcs11session.c (hunk 1 above)
Comment (gamringer) [reply pkcs11session.c]: It should return a CK_RV

**Comment by gamringer** [comment]
I'm working on a few changes

**Comment by gamringer** [comment]
@JanSlabon I agree, if the RFC calls it URI, then we should also call it such.

Regarding the method name, the idea was to have a Pkcs11\Uri class with an open(string $uri) method, This method would then return the thing to which the uri points, which could be a module, an open session, or an object. Therefore, to keep things consistent, the corresponding method on the Pkcs11\Session class would be openUri(string $uri), but would only return objects retrieved from the current open session.

**Comment by gamringer** [comment]
@JanSlabon @JimmyHamel @vjardin let me know what you think with this latest commit.

**Comment by JimmyHamel** [comment]
I agree on the url to uri change. The latest looks good, just removed a commented error log.

**Comment by JanSlabon** [comment]
> @JanSlabon @JimmyHamel @vjardin let me know what you think with this latest commit.

I was on vacation, so sorry for the late feedback :-) If the method name was intentional I'm fine with it.

**Pull request merged**